### ISSUE TITLE
Support Half-precision floating-point numbers

### ISSUE THREAD

**Issue by vermorel** [dotnet/coreclr#11948]
The CLR would greatly benefit from adding support the half-precision floating-point numbers.

The [Half-precision floating-point numbers](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) are already extensively used in machine learning and image processing applications. Even if it the CLR is only meant for interacting with a native machine learning framework, the lack of a `System.Half` number means that those needs to be emulated, which is very inefficient; especially considering that both ARM and x64 have built-in support for half-precision floating point numbers.

Emulation can be done ala [C# Half](https://sourceforge.net/projects/csharp-half/) for example, but my own experiments hinted about 10x slower compared to native performance. Then, I wasn't even using dedicated SIMD, which are likely to add another 10x performance factor (bringing to 100x) between "emulated" and "native" on high-end processors.

**Comment by fanoI** [comment]
Related to https://github.com/dotnet/corefx/issues/17267 (called binary16 there).

I don't know what is needed to make these things move... I think this experimental work should be done in corefxlab.

**Comment by danmosemsft** [comment]
@vermorel if you agree it makes sense, let's close this in favor of  dotnet/corefx#17267. Perhaps you can update that with your thoughts. The experts to ask for feedback on it are mellinoe and tannergooding. They can help determine whether it's worth doing and if so what the API would look like, etc. 

We use CoreFX for tracking API additions: if this manifests with API chagnes, it's probably the best place to track even if the implementation would involve CLR work.

**Comment by vermorel** [comment]
@danmosemsft @tannergooding @mellinoe  I might disagree with dotnet/corefx#17267

More specifically, not all numerical types are equal. The `binary16` (IEEE name) is vastly more important because it comes with massive hardware support both in CPUs and GPUs. In contrast, `binary128` have close to zero hardware support (the defunct Itanium), and would be best left emulated. Idem for `BigDecimal`.

Then, the half-precision should be named `System.Half` and located in the `System` namespace. The name `Half` is consistent with `System.Single` and `System.Double` that exist in the CLR. This floating point number should _not_ be treated as an advanced numerical type; on the contrary it's a fairly basic type. Then, it happens that `Vector<Half>` would be a great addition, but that's another topic altogether.

What do you think?

**Comment by tannergooding** [comment]
> comes with massive hardware support both in CPUs and GPUs
It certainly has GPU support, but I think that is less important given that the runtime doesn't support executing on GPU.

CPU support is also very limited, Intel only provides two instructions that work with half-precision values: `VCVTPH2PS` and `VCVTPS2PH`. These instructions are only used for converting between half and single-precision values.

The benefit of `System.Half` would be that it could be "fast" if all operations were implemented as conversion to `System.Single`, computed, and then case back. However, this has the same caveats of `System.Single`->`System.Double`, compute, and `System.Double`->`System.Single` in that results in some scenarios could be vastly different then if they were computed entirely as `System.Single` (due to the extended precision preventing overflow/underflow).

Additionally, even if `System.Half` did provide complete hardware support on one platform, the other platforms we support might not have equivalent support, so software emulation would be required at some level either way.

I imagine the best implementation would be similar to `System.Numerics.Vector`, where we emulate if there is no hardware support; otherwise, we just compile down to an intrinsic.

**Comment by vermorel** [comment]
@tannergooding  I agree to your point that hardware support will be heterogeneous at best, with emulation needed in many situations. Yet the same can be said for `System.Decimal`. Then, if we look at the industry trends, hardware support for smaller data types is bound to increase, and that's exactly has been happening over the last couple of years. Conversely, the _bigger_ data types have followed the opposite direction, with Itanium and its `binary128` support being terminated, and with AFAIK no plans to bring them back.

Then, I don't exactly agree on the fact that because CLR does not execute on GPU that it makes `binary16` any less important.  Any .NET app that involves machine learning is likely to have `binary16` values at the boundaries, typically through interop to move data to a dedicated framework like [Caffe2](https://caffe2.ai/blog/2017/05/10/caffe2-adds-FP16-training-support.html).  Then, even GPUs are coming to .NET through tools like [Quantalea](http://www.quantalea.com/features/). A first-class `binary16` would streamline the boundaries.

**Comment by tannergooding** [comment]
@vermorel, I had meant that GPU support is less important since the runtime itself doesn't support executing on it, not that `binary16` was less important 😄. That is, ease of implementation for `System.Half` (as well as any optimizations that the runtime can provide) comes down to the architectures we currently support (with some consideration to future platforms we may support).

I do definitely agree that `binary16` would be easier to implement than any of the other `binary*` types and would likely find more use, making it a better candidate type if we were to decide whether or not to implement the full set of `IEEE` types.

I don't have much preference for what namespace the type exist under either, as I would likely be doing `using half = <namespace>.Half` most of the time anyways.

**Comment by tannergooding** [comment]
https://github.com/dotnet/designs/issues/13

**Comment by vermorel** [comment]
FYI, CNTK (deep learning at Microsoft) will be adding `binary16` to their framework. See their [August 2017 roadmap](https://github.com/Microsoft/CNTK/issues/2194). It would also complement their .NET plans.

**Comment by danmosemsft** [comment]
@ericsj for comment above about CNTK

**Comment by ericstj** [comment]
We technically could interop with CNTK with a struct emulation.  With CNTK all the heavy lifting will happen in the native code side of things so perf of the managed representation isn't critical, at least for evaluation.  A proper type will help more when you start thinking about .NET training scenarios where more work is happening on the .NET side.  Regardless I see the value of this addition and think it's something we should do.
